torch/cuda/__init__.py

```diff
@@ -544,6 +544,15 @@ def _device_count_nvml() -> int:
         return raw_cnt
     return len(set(range(raw_cnt)).intersection(visible_devices))
 
+def _get_nvml_device_index(device: Optional[Union[int, Device]]) -> int:
+    r"""Returns the NVML index of the device, taking CUDA_VISIBLE_DEVICES into account."""
+    idx = _get_device_index(device, optional=True)
+    visible_devices = _parse_visible_devices()
+    idx_map = {idx: real_idx for idx, real_idx in enumerate(sorted(visible_devices))}
```

Comment:
The `visible_devices` should not be sorted.

Reply:
It's a `set`, it has to be sorted or the order is arbitrary

Reply:
It shouldn't be a `set`. The implementation of `_parse_visible_devices` is wrong. `CUDA_VISIBLE_DEVICES='0,1'` and `CUDA_VISIBLE_DEVICES='1,0'` have different meanings.

Reply:
Ok, I assumed that was normalized in by the C++ code or CUDA libraries. I can refactor that function to return a list with the order preserved if that's not the case